File: packages/mobile/scripts/backfill_discoveries.py
#!/usr/bin/env python3
"""
Local backfill script to process pending discoveries

Usage:
    python3 scripts/backfill_discoveries.py [--limit N] [--delay SECONDS]

Examples:
    python3 scripts/backfill_discoveries.py                  # Process all
    python3 scripts/backfill_discoveries.py --limit 10       # Process 10
    python3 scripts/backfill_discoveries.py --limit 5 --delay 3  # Process 5 with 3s delay
"""

import requests
import json
import time
import sys
from datetime import datetime

SUPABASE_URL = "https://nidbhgqeyhrudtnizaya.supabase.co"
SERVICE_ROLE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Im5pZGJoZ3FleWhydWR0bml6YXlhIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc1OTM4OTMwNywiZXhwIjoyMDc0OTY1MzA3fQ.Ae1SOpALHEq0K68a0cwK38ugbHx4hKiqzC28q1Hkf6M"

# Parse command line args
limit = None
delay = 5

for i, arg in enumerate(sys.argv):
    if arg == "--limit" and i + 1 < len(sys.argv):
        limit = int(sys.argv[i + 1])
    elif arg == "--delay" and i + 1 < len(sys.argv):
        delay = int(sys.argv[i + 1])

print("🚀 Starting local backfill...")
print(f"   Delay: {delay}s between requests")
if limit:
    print(f"   Limit: {limit} discoveries")
print()

# Fetch pending discoveries
print("📊 Fetching pending discoveries...")

headers = {
    "apikey": SERVICE_ROLE_KEY,
    "Authorization": f"Bearer {SERVICE_ROLE_KEY}",
    "Content-Type": "application/json"
}

# Build query
url = f"{SUPABASE_URL}/rest/v1/discovered_listings"
params = {
    "status": "eq.pending_research",
    "order": "engagement_score.desc",
    "select": "*"
}
if limit:
    params["limit"] = limit

response = requests.get(url, headers=headers, params=params)

if response.status_code != 200:
    print(f"❌ Error fetching discoveries: {response.status_code}")
    print(response.text)
    sys.exit(1)

discoveries = response.json()

if not discoveries:
    print("✅ No pending discoveries found!")
    sys.exit(0)

print(f"Found {len(discoveries)} pending discoveries\n")

# Statistics
processed = 0
succeeded = 0
failed = 0
start_time = time.time()

# Process each discovery
for i, discovery in enumerate(discoveries):
    processed += 1
    progress = f"[{processed}/{len(discoveries)}]"

    name = discovery.get('name', 'Unknown')
    location = discovery.get('location', 'Unknown')
    city = discovery.get('city', 'Unknown')
    state = discovery.get('state', 'Unknown')
    service_type = discovery.get('service_type') or 'venue'
    engagement = discovery.get('engagement_score', 0)

    print(f"{progress} 🔍 Researching: {name}")
    print(f"   📍 {location}, {city}")
    print(f"   📊 Engagement: {engagement}/10")
    print(f"   🏷️  Type: {service_type}")

    try:
        # Call deep-research-venue Edge Function
        payload = {
            "venueName": name,
            "location": f"{location}, {city}",
            "city": city,
            "state": state,
            "serviceType": service_type,
            "forceRefresh": False
        }

        research_response = requests.post(
            f"{SUPABASE_URL}/functions/v1/deep-research-venue",
            headers=headers,
            json=payload,
            timeout=60
        )

        if research_response.status_code != 200:
            raise Exception(f"HTTP {research_response.status_code}: {research_response.text}")

        result = research_response.json()

        if result.get('success'):
            succeeded += 1
            listing = result.get('listing', {})
            listing_id = listing.get('id')
            photos = len(listing.get('images', []))
            packages = len(listing.get('packages', []))

            print(f"   ✅ Success! Created listing: {listing_id}")
            print(f"   📸 Photos: {photos}")
            print(f"   📦 Packages: {packages}")

            # Update discovery status
            update_response = requests.patch(
                f"{SUPABASE_URL}/rest/v1/discovered_listings?id=eq.{discovery['id']}",
                headers={**headers, "Prefer": "return=minimal"},
                json={
                    "status": "researched",
                    "listing_id": listing_id,
                    "researched_at": datetime.utcnow().isoformat()
                }
            )

        else:
            failed += 1
            error_msg = result.get('error', 'Unknown error')
            print(f"   ❌ Failed: {error_msg}")

            # Mark as failed
            requests.patch(
                f"{SUPABASE_URL}/rest/v1/discovered_listings?id=eq.{discovery['id']}",
                headers={**headers, "Prefer": "return=minimal"},
                json={"status": "research_failed"}
            )

    except Exception as e:
        failed += 1
        print(f"   ❌ Error: {str(e)}")

        # Mark as failed
        try:
            requests.patch(
                f"{SUPABASE_URL}/rest/v1/discovered_listings?id=eq.{discovery['id']}",
                headers={**headers, "Prefer": "return=minimal"},
                json={"status": "research_failed"}
            )
        except:
            pass

    # Rate limiting
    if processed < len(discoveries):
        print(f"   ⏱️  Waiting {delay}s...")
        time.sleep(delay)

    # Progress summary every 10 items
    if processed % 10 == 0 or processed == len(discoveries):
        elapsed = time.time() - start_time
        rate = processed / elapsed if elapsed > 0 else 0
        remaining = len(discoveries) - processed
        eta = remaining / rate if rate > 0 else 0

        print()
        print("📈 Progress Summary:")
        print(f"   Processed: {processed}/{len(discoveries)}")
        print(f"   ✅ Succeeded: {succeeded}")
        print(f"   ❌ Failed: {failed}")
        print(f"   ⏱️  Elapsed: {int(elapsed)}s")
        print(f"   🚀 Rate: {rate:.2f}/s")
        if remaining > 0:
            print(f"   ⏰ ETA: {int(eta)}s ({int(eta/60)}min)")

    print()

# Final summary
total_time = time.time() - start_time

print("=" * 60)
print("🎉 Backfill Complete!")
print("=" * 60)
print(f"Total Processed: {processed}")
print(f"✅ Succeeded: {succeeded} ({succeeded/processed*100:.0f}%)")
print(f"❌ Failed: {failed} ({failed/processed*100:.0f}%)")
print(f"⏱️  Total Time: {int(total_time)}s ({int(total_time/60)}min)")
print(f"🚀 Average Rate: {processed/total_time:.2f}/s")
print()

# Fetch updated stats
stats_response = requests.get(
    f"{SUPABASE_URL}/rest/v1/discovered_listings",
    headers=headers,
    params={"select": "status"}
)

if stats_response.status_code == 200:
    stats = stats_response.json()
    status_counts = {}
    for item in stats:
        status = item.get('status', 'unknown')
        status_counts[status] = status_counts.get(status, 0) + 1

    print("📊 Updated Discovery Queue:")
    print(f"   Pending: {status_counts.get('pending_research', 0)}")
    print(f"   Researched: {status_counts.get('researched', 0)}")
    print(f"   Failed: {status_counts.get('research_failed', 0)}")

# Count total listings
listings_response = requests.get(
    f"{SUPABASE_URL}/rest/v1/normalized_listings",
    headers={**headers, "Prefer": "count=exact"},
    params={"select": "id", "limit": 1}
)

if listings_response.status_code == 200:
    total_listings = listings_response.headers.get('Content-Range', '').split('/')[-1]
    if total_listings:
        print(f"\n🏆 Total Listings in Database: {total_listings}")

File: packages/mobile/scripts/backfill-discoveries.ts
#!/usr/bin/env -S deno run --allow-net --allow-env

/**
 * Local backfill script to process pending discoveries
 *
 * Usage:
 *   deno run --allow-net --allow-env scripts/backfill-discoveries.ts
 *
 * Options:
 *   --limit=N     Process only N discoveries (default: all)
 *   --chunk=N     Process N at a time (default: 10)
 *   --delay=MS    Delay between requests in ms (default: 5000)
 */

import { createClient } from 'https://esm.sh/@supabase/supabase-js@2'

const SUPABASE_URL = 'https://nidbhgqeyhrudtnizaya.supabase.co'
const SUPABASE_SERVICE_ROLE_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Im5pZGJoZ3FleWhydWR0bml6YXlhIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc1OTM4OTMwNywiZXhwIjoyMDc0OTY1MzA3fQ.Ae1SOpALHEq0K68a0cwK38ugbHx4hKiqzC28q1Hkf6M'

// Parse command line args
const args = Deno.args.reduce((acc, arg) => {
  if (arg.startsWith('--')) {
    const [key, value] = arg.slice(2).split('=')
    acc[key] = value || true
  }
  return acc
}, {} as Record<string, any>)

const LIMIT = args.limit ? parseInt(args.limit) : undefined
const CHUNK_SIZE = args.chunk ? parseInt(args.chunk) : 10
const DELAY_MS = args.delay ? parseInt(args.delay) : 5000

console.log('🚀 Starting local backfill...')
console.log(`   Chunk size: ${CHUNK_SIZE}`)
console.log(`   Delay: ${DELAY_MS}ms`)
if (LIMIT) console.log(`   Limit: ${LIMIT} discoveries`)
console.log('')

const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY)

// Fetch pending discoveries
console.log('📊 Fetching pending discoveries...')
let query = supabase
  .from('discovered_listings')
  .select('*')
  .eq('status', 'pending_research')
  .order('engagement_score', { ascending: false })

if (LIMIT) {
  query = query.limit(LIMIT)
}

const { data: discoveries, error } = await query

if (error) {
  console.error('❌ Error fetching discoveries:', error)
  Deno.exit(1)
}

if (!discoveries || discoveries.length === 0) {
  console.log('✅ No pending discoveries found!')
  Deno.exit(0)
}

console.log(`Found ${discoveries.length} pending discoveries\n`)

// Statistics
let processed = 0
let succeeded = 0
let failed = 0
let skipped = 0

const startTime = Date.now()

// Process in chunks
for (let i = 0; i < discoveries.length; i += CHUNK_SIZE) {
  const chunk = discoveries.slice(i, Math.min(i + CHUNK_SIZE, discoveries.length))
  const chunkNum = Math.floor(i / CHUNK_SIZE) + 1
  const totalChunks = Math.ceil(discoveries.length / CHUNK_SIZE)

  console.log(`\n📦 Chunk ${chunkNum}/${totalChunks} (${chunk.length} discoveries)`)
  console.log('─'.repeat(60))

  for (const discovery of chunk) {
    processed++
    const progress = `[${processed}/${discoveries.length}]`

    try {
      console.log(`${progress} 🔍 Researching: ${discovery.name}`)
      console.log(`   📍 ${discovery.location}, ${discovery.city}`)
      console.log(`   📊 Engagement: ${discovery.engagement_score}/10`)
      console.log(`   🏷️  Type: ${discovery.service_type || 'venue'}`)

      // Call the deep-research-venue Edge Function
      const response = await fetch(`${SUPABASE_URL}/functions/v1/deep-research-venue`, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${SUPABASE_SERVICE_ROLE_KEY}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          venueName: discovery.name,
          location: `${discovery.location}, ${discovery.city}`,
          city: discovery.city,
          state: discovery.state,
          serviceType: discovery.service_type || 'venue',
          forceRefresh: false
        })
      })

      if (!response.ok) {
        const errorText = await response.text()
        throw new Error(`HTTP ${response.status}: ${errorText}`)
      }

      const result = await response.json()

      if (result.success) {
        succeeded++
        console.log(`   ✅ Success! Created listing: ${result.listing.id}`)
        console.log(`   📸 Photos: ${result.listing.images?.length || 0}`)
        console.log(`   📦 Packages: ${result.listing.packages?.length || 0}`)

        // Update discovery status
        await supabase
          .from('discovered_listings')
          .update({
            status: 'researched',
            listing_id: result.listing.id,
            researched_at: new Date().toISOString()
          })
          .eq('id', discovery.id)

      } else {
        failed++
        console.log(`   ❌ Failed: ${result.error || 'Unknown error'}`)

        // Mark as failed
        await supabase
          .from('discovered_listings')
          .update({ status: 'research_failed' })
          .eq('id', discovery.id)
      }

    } catch (error) {
      failed++
      console.log(`   ❌ Error: ${error.message}`)

      // Mark as failed
      await supabase
        .from('discovered_listings')
        .update({
          status: 'research_failed',
          error_message: error.message
        })
        .eq('id', discovery.id)
    }

    // Rate limiting delay between requests
    if (processed < discoveries.length) {
      console.log(`   ⏱️  Waiting ${DELAY_MS/1000}s...`)
      await new Promise(resolve => setTimeout(resolve, DELAY_MS))
    }
  }

  // Progress summary after each chunk
  const elapsed = (Date.now() - startTime) / 1000
  const rate = processed / elapsed
  const remaining = discoveries.length - processed
  const eta = remaining / rate

  console.log('\n📈 Progress Summary:')
  console.log(`   Processed: ${processed}/${discoveries.length}`)
  console.log(`   ✅ Succeeded: ${succeeded}`)
  console.log(`   ❌ Failed: ${failed}`)
  console.log(`   ⏱️  Elapsed: ${Math.round(elapsed)}s`)
  console.log(`   🚀 Rate: ${rate.toFixed(2)}/s`)
  if (remaining > 0) {
    console.log(`   ⏰ ETA: ${Math.round(eta)}s (${Math.round(eta/60)}min)`)
  }
}

// Final summary
const totalTime = (Date.now() - startTime) / 1000

console.log('\n' + '='.repeat(60))
console.log('🎉 Backfill Complete!')
console.log('='.repeat(60))
console.log(`Total Processed: ${processed}`)
console.log(`✅ Succeeded: ${succeeded} (${Math.round(succeeded/processed*100)}%)`)
console.log(`❌ Failed: ${failed} (${Math.round(failed/processed*100)}%)`)
console.log(`⏱️  Total Time: ${Math.round(totalTime)}s (${Math.round(totalTime/60)}min)`)
console.log(`🚀 Average Rate: ${(processed/totalTime).toFixed(2)}/s`)
console.log('')

// Fetch updated stats
const { data: stats } = await supabase
  .from('discovered_listings')
  .select('status')

if (stats) {
  const statusCounts = stats.reduce((acc: any, d: any) => {
    acc[d.status] = (acc[d.status] || 0) + 1
    return acc
  }, {})

  console.log('📊 Updated Discovery Queue:')
  console.log(`   Pending: ${statusCounts.pending_research || 0}`)
  console.log(`   Researched: ${statusCounts.researched || 0}`)
  console.log(`   Failed: ${statusCounts.research_failed || 0}`)
}

const { data: listingsCount } = await supabase
  .from('normalized_listings')
  .select('id', { count: 'exact', head: true })

if (listingsCount) {
  console.log(`\n🏆 Total Listings in Database: ${listingsCount.length || 0}`)
}
